keps/sig-apps/4368-support-managed-by-label-for-batch-jobs/README.md

```diff
@@ -0,0 +1,1193 @@
+
+#### Label mutability
+
+We keep the label immutable. See also the discussion in
+[Can the label be mutable?](#can-the-label-be-mutable).
+
+### Test Plan
+
+[x] I/we understand the owners of the involved components may require updates to
+existing tests to make this code solid enough prior to committing the changes necessary
+to implement this enhancement.
+
+##### Prerequisite testing updates
+
+##### Unit tests
+
+- `pkg/controller/job`: `2023-12-20` - `91.5%`
+- `pkg/registry/batch/job`: `2023-12-20` - `92.2%`
+- `pkg/apis/batch/validation`: `2024-01-22` - `98.0%`
+- `pkg/apis/batch/v1`: `2023-12-20` - `29.3%` (mostly generated code)
+
+The following scenarios are covered:
+- the Job controller reconciles jobs with the "managed-by" label equal to `job-controller.k8s.io` when the feature is enabled
+- the Job controller reconciles jobs without the "managed-by" label when the feature is enabled
+- the Job controller does not reconcile jobs with custom value of the "managed-by" label when the feature is enabled
+- the Job controller reconciles jobs with custom "managed-by" label when the feature gate is disabled
+- verify the label is immutable, both when the job is suspended or unsuspended; when the feature is enabled
+- enablement / disablement of the feature after the Job (with custom "managed-by" label) is created
+- verify the new Job Status API validation rules (see [here](#job-status-validation))
+
+##### Integration tests
+
+The following scenarios are covered:
+- the Job controller reconciles jobs with the "managed-by" label equal to `job-controller.k8s.io`
+- the Job controller reconciles jobs without the "managed-by" label
+- the Job controller does not reconcile a job with any other value of the "managed-by" label. In particular:
+  - it does not reset the status for a Job with `.spec.suspend=false`,
+  - it does not add the Suspended condition for a Job with `.spec.suspend=true`.
+- the Job controller reconciles jobs with custom "managed-by" label when the feature gate is disabled
+- the `job_by_external_controller_total` metric is incremented when a new Job with custom "managed-by" is created
+- the `job_by_external_controller_total` metric is not incremented for a new Job without "managed-by" or with default value
+- the `job_by_external_controller_total` metric is not incremented for Job updates (regardless of the "managed-by")
+
+During the implementation more scenarios might be covered.
+
+##### e2e tests
+
+The feature does not depend on kubelet, so the functionality can be fully
+covered with unit & integration tests.
+
+We propose a single e2e test for the following scenario:
+- the Job controller does not reconcile a job with any other value of the "managed-by" label. In particular,
+  it does not reset the status for an unsuspended Job.
+
+### Graduation Criteria
+
+#### Beta
+
+- skip synchronization of jobs when the "managed-by" label does not exist, or equals `job-controller.k8s.io`
+- unit & integration tests
```

Comment:
With beta enabled by default, e2e tests should also be in beta criteria 
  [i.e. we shouldn't be by default giving to users something that we don't test e2e]

Reply:
Good point, updated. PTAL.